Unify list and decomposed metric interfaces

sushobhan2024 opened the pull request:
- [x] Moved decomposed metric methods into the main `Metric` class, renamed for clarity
- [x] Adjusted `summarize` method interface for broader input support
- [x] Added default `summarize` and `extract_list_metrics` implementations in `ListMetric`
- [x] Kept `DecomposedMetric` as deprecated wrapper for compatibility
- [x] Added a class `MetricAccumulator` that includes `add_metric`, `measure_list`, `list_metrics`, and `summary_metrics`.
- [x] Modify `RunAnalysis` to use `MetricAccumulator`.
- [x] Write tests.

mdekstrand commented on src/lenskit/metrics/_base.py:
```diff
@@ -43,6 +46,34 @@ def label(self) -> str:
+    @abstractmethod
+    def measure_list(self, output: ItemList, test: ItemList, /) -> object:
+        """
+        Compute measurements for a single list.
```
We should define what the return data is and how it is intended to be used, and why it might be something other than a `float` or `dict`.

sushobhan2024 replied:
I have updated with the return data and included `float` and `dict`. What other data types should we allow for `measure_list` return?

mdekstrand commented on src/lenskit/metrics/_base.py:
```diff
@@ -95,6 +145,11 @@ def measure_run(self, output: ItemListCollection, test: ItemListCollection, /) -
 class DecomposedMetric(Metric):
     """
+    Deprecated base class for decomposed metrics.
+    .. deprecated:: 2025.4
+        This class is deprecated and its functionality has been moved to :class:`Metric`.
+        It is scheduled for removal in 2026.
```
I think we need a blank line before and after `.. deprecated`.

sushobhan2024 commented on src/lenskit/metrics/_base.py:
```diff
@@ -72,6 +103,25 @@ def measure_list(self, output: ItemList, test: ItemList, /) -> float:
+
+    def summarize(self, values: list[object] | np.ndarray | pa.Array | pa.ChunkedArray, /) -> float:
+        """
+        Aggregarte per-list metric values by computing their mean, ignoring NaNs.
```
I have now removed the docstring from the method implementation, and added a note in the class's docstring.

mdekstrand commented on src/lenskit/metrics/_accum.py:
```diff
@@ -0,0 +1,77 @@
+        self,
+        output: ItemList,
+        test: ItemList,
+        user_id: Optional[Any] = None,
```
This should be a kwargs of keys — not all evaluations have user IDs, and some may have more than one key. The keys should work like `ItemListCollection` (in fact, the same tuple type logic can be used).

mdekstrand commented on src/lenskit/metrics/_accum.py:
```diff
@@ -0,0 +1,77 @@
+        """
+        record = {"user_id": user_id}
+        for metric in self._metrics:
+            result = metric.measure_list(output, test)
```
We need to keep the results around too — the aggregate will need them.

mdekstrand commented on src/lenskit/metrics/_accum.py:
```diff
@@ -0,0 +1,77 @@
+        summaries = {}
+        df = self.list_metrics()
+        for metric in self._metrics:
+            values = df[metric.label].dropna().tolist()
```
I don't think we should filter data here — if we are to drop NAs, the summarizer should do it. I also don't think we want to convert to `list` — those are inefficient (this is where we may need a little bit of a rethink on the types on the metric methods).

mdekstrand replied:
Also, the summarizer should get the list measurement results, not the extracted metrics.

For most of metrics currently in LensKit, there is no difference. However, in the future, there will be: the intermediate results may not be metrics. This is also the case for RMSE: the per-list result for RMSE is the sum of squares and the number of ratings. Extracting a list metric converts those into an RMSE; computing the global micro-averaged RMSE sums them up to compute a global RMSE.

mdekstrand commented on src/lenskit/metrics/_accum.py:
```diff
@@ -0,0 +1,77 @@
+        record = {"user_id": user_id}
+        for metric in self._metrics:
+            result = metric.measure_list(output, test)
+            record[metric.label] = metric.extract_list_metrics(result)
```
Is this what we want if the list metrics return a dict? Possibly, but need to take a think about it.

mdekstrand commented on src/lenskit/metrics/_accum.py:
```diff
@@ -0,0 +1,77 @@
+from ._base import Metric
+
+
+class MetricAccumulator:
```
We need tests for this class.

mdekstrand commented on src/lenskit/metrics/_accum.py:
```diff
@@ -0,0 +1,77 @@
+        Compute overall summary statistics by aggregating per-list metrics.
+
+        Returns:
+            A dictionary mapping metric labels to their aggregated summary values.
```
What if a metric produces more than one summary value?

mdekstrand reviewed (changes requested):
The API for `MetricAccumulator` looks pretty good. I was pondering whether it should return an analysis result object like `RecListAnalysis` currently does; I am not sure. What do you think? We will need to create analysis results for compatibility when we switch `RecListAnalysis` to use this code.

We do need some revision on the list ID API (it can't just be a user ID), and on how we're handling metric results (both the intermediate results from `measure_list`, and the extracted and summarized metrics). Two key points there:

- We need to save intermediate results, because summarization works on those, not on the extracted metrics.
- Both `extract_list_metrics` and `summarize` can return more than one metric value, and we need to make sure that we handle these the way we want to, and document how they are handled. Nested dictionaries can work, but quickly get annoying and take a lot of space. We might want instead to use dot-separated names: if a metric "RecipRank" produces a summary named "mean", that output value becomes "RecipRank.mean".

sushobhan2024 commented:
@mdekstrand Yes, I think returning an object like `RunAnalysisResults` from the accumulator provides a consistent way to access both per-list and summary metrics. In this update, I have:

1. added `to_result` method that returns `RunAnalysisResults` object. 
2. added `normalize_metrics` method to flatten dictionary in case of more than one summary values. `extract_list_metrics` and `summarize` use this method.
3. refactored `RunAnalysis.measure` to used `MetricAccumulator`, and
4. added tests for the accumulator

I was thinking if intermediate results are needed anywhere else other than in summarization. Do we need to add a method to return intermediate results as well?

Also, should the summary include only computed metrics, or also registered but unmeasured ones? After adding a test for unmeasured registered metrics, I added `defaults` to the `to_result` method so they are included in the summary.

mdekstrand reviewed (commented):
Need to get tests passing, but the design looks good to me now.

sushobhan2024 commented on src/lenskit/metrics/_accum.py:
```diff
@@ -0,0 +1,124 @@
+        )
+
+
+def normalize_metrics(label: str, value: float | dict[str, float] | None) -> dict[str, float]:
```
I agree `normalize_metrics` is a bit generic. I will rename it to `to_metric_dict`.

mdekstrand commented:
@sushobhan2024 yes it should.

mdekstrand commented:
I suspect the evaluation-based test is a change in behavior in the RMSE metric.

sushobhan2024 commented on src/lenskit/metrics/_accum.py:
```diff
@@ -0,0 +1,301 @@
+            if isinstance(m, ListMetric):
+                default = m.default
+            else:
+                default = 0.0
```
I think I copy-pasted this from the old codebase. Are you asking me to change `ListMetric` to `Metric` in the `isinstance` check, or are you suggesting I should write a test case to verify this logic?

mdekstrand replied:
My question is specifically about defaulting the default to 0. There are many cases in which that would not be correct.

Dealing with defaults is complex and subtle, and may actually be more complexity than is good to have. With just a list metric, it was easy, but with a more complex notion of metric outputs, it's less straightforward. However, metrics are also now responsible for their own summarization, which helps us — the primary purpose of the default was to correctly input something into the summarization process.

Therefore, we might actually want to pull the use of the `default` logic into the `ListMetric` implementation of `summarize`, and make a note for LensKit 2026 to make metrics directly return their default values when appropriate. We have a period we need to keep some compatibility, but we might want to yeet this concept of “defaults” as a metric property.

sushobhan2024 commented on src/lenskit/metrics/_accum.py:
```diff
@@ -0,0 +1,301 @@
+
+        for wrapper in self.metrics:
+            try:
+                if wrapper.is_decomposed:
+                    data = wrapper.compute_list_data(output, test)
+                    self._list_data[wrapper.label].append(data)
+                    list_metric = wrapper.extract_list_metrics(data)
+                    self._list_metrics[wrapper.label].append(list_metric)
+                else:
+                    metric_value = wrapper.measure_list(output, test)
+                    self._list_data[wrapper.label].append(metric_value)
+                    self._list_metrics[wrapper.label].append(metric_value)
```
You're right. I have now simplified this. For the global metrics, should we keep the logic in RunAnalysis as in the old code, or add a method in the accumulator?

sushobhan2024 commented on src/lenskit/metrics/bulk.py:
```diff
@@ -238,68 +178,37 @@ def measure(
         if no_test_count:
             _log.warning("could not find test data for %d lists", no_test_count)
 
-        _log.debug("computing %d global metrics for %d output lists", len(gms), n)
-        global_results = pd.Series(
-            {
-                m.label: (
-                    m.metric.global_aggregate(list_intermediates[m.label])
-                    if m.is_decomposed
-                    else m.measure_run(outputs, test)
-                )
-                for m in self.metrics
-                if m.is_global or m.is_decomposed
-            },
-            dtype=np.float64,
-        )
```
I realize I have been assigning this to `summarize` instead which is incorrect.

mdekstrand commented on src/lenskit/metrics/_base.py:
```diff
@@ -22,27 +25,79 @@ class Metric(ABC):
+                elif isinstance(v, dict) and "mean" in v:
+                    numeric_values.append(float(v["mean"]))
+
+        if not numeric_values:
+            return {"mean": 0.0}
```
We should return `None` for empty results.

mdekstrand commented on src/lenskit/metrics/_accum.py:
```diff
@@ -38,8 +38,10 @@ class MetricWrapper:
-        return isinstance(self.metric, (ListMetric, Callable))
+        "Check if this metric is listwise (includes ListMetric and callable functions)."
+        return isinstance(self.metric, (ListMetric, Callable)) or (
+            isinstance(self.metric, Metric) and hasattr(self.metric, "measure_list")
```
we shouldn't need both of these tests — `measure_list` is defined by `Metric`.

mdekstrand replied:
Maybe this whole test should just test that it does not implement `GlobalMetric`? (I don't remember the exact relationship between `GlobalMetric` and `DecomposedMetric` right now.)

sushobhan2024 replied:
The accumulator code treats anything that is not `GlobalMetric` as list wise. Should we can use a negative test, something like `not isinstance(self.metric, GlobalMetric)`, or use the code as in the old codebase? I have reverted to the old code now.

There were 2 eval tests failing after I pushed the code. However, running the tests in my computer skipped those tests. How do we address this?

mdekstrand commented:
@sushobhan2024 This is the important bit:
> ```
> error: rustc 1.86.0 is not supported by the following package:
>    nalgebra@0.34.0 requires rustc 1.87.0
> ```

Run `rustup update` to get a newer Rust.

mdekstrand commented on src/lenskit/metrics/_accum.py:
```diff
@@ -0,0 +1,295 @@
+    def extract_list_metrics(self, intermediate_data: Any) -> float | dict[str, float] | None:
+        """Extract per-list metrics from intermediate data."""
+        result = self.metric.extract_list_metrics(intermediate_data)
+        return intermediate_data if result is None else result
```
This should just return `result` — if the metric doesn't return a value from `extract_list_metrics`, then it doesn't support list metrics. We can't assume that the intermediate data is list metrics.

mdekstrand commented on src/lenskit/metrics/_accum.py:
```diff
@@ -0,0 +1,295 @@
+            return result
+        return {"mean": float(result), "median": None, "std": None}
+
+    def _default_summarize(self, values) -> dict[str, float | None]:
+        """A fallback for metrics that do not implement their own summarize method."""
+        if isinstance(values, (pa.Array, pa.ChunkedArray)):
+            values = values.to_pylist()
+        numeric_values = [v for v in values if v is not None]
+
+        if not numeric_values:
+            return {"mean": None, "median": None, "std": None}
+
+        arr = np.array(numeric_values, dtype=np.float64)
+        return {
+            "mean": float(np.mean(arr)),
+            "median": float(np.median(arr)),
+            "std": float(np.std(arr, ddof=1)) if len(arr) > 1 else 0.0,
+        }
```
I don't think we need default summarization — `ListMetric` implements default summarization logic, and for other metrics we can't assume that it is meaningful to compute summary statistics over intermediate data. If we did provide default summarization, it would need to be over the extracted list metrics, and I think we should defer that to a future enhancement if it's even something we want.

mdekstrand commented on src/lenskit/metrics/_accum.py:
```diff
@@ -0,0 +1,295 @@
+            return self._default_summarize(values)
+        if isinstance(result, dict):
+            return result
+        return {"mean": float(result), "median": None, "std": None}
```
We can't assume that a single result is a mean — it could just be a single summary value. For example, Gini will return a single value.

mdekstrand commented on src/lenskit/metrics/_base.py:
```diff
@@ -22,27 +25,76 @@ class Metric(ABC):
+
+    default: ClassVar[float | None] = 0.0
+    """
+    The default value to infer when computing statistics over missing values.
+    If ``None``, no inference is done (necessary for metrics like RMSE, where
+    the missing value is theoretically infinite).
```
"infer" -> "impute", "inference" -> "imputation".

mdekstrand replied:
Should `default` live on `ListMetric` instead of `Metric`? A single `default` value for a metric, that may compute more than one metric, doesn't seem to make sense.

mdekstrand reviewed (changes requested):
Almost there! We just need to drop some excess logic that assumes more about metrics than the interface guarantees, and I think we'll be ready to merge.

sushobhan2024 commented on src/lenskit/metrics/ranking/_gini.py:
```diff
@@ -50,24 +50,11 @@ class ListGini(GiniBase):
 
     This computes the Gini coefficient of the *number of lists* that each item
     appears in.
-
-    Args:
-        k:
-            The maximum recommendation list length.
-        items:
-            The total number of items, a data frame or series of item data, or a
-            dataset. If a frame or series is provided, its length will be used
-            as the number of items.  If a dataset is provided, its item count
-            will be used.
-
-    Stability:
-        Caller
```
Will keep this. This should be a mistake. I don't remember removing.
